[BEAM-6892] Adding support for side inputs on table & schema. Also adding additional params for BQ.

pabloem opened the pull request:
r: @chamikaramj 
This will allow support for time partitioning, and for side inputs coming into the schema and table transforms. This is useful because schemas or table names may be computed during pipeline execution, and side inputs allow users the ability to access them when they're computed.

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
This is available in Java via DynamicDestination class: https://beam.apache.org/releases/javadoc/2.11.0/org/apache/beam/sdk/io/gcp/bigquery/DynamicDestinations.html

chamikaramj commented on sdks/python/apache_beam/io/gcp/bigquery.py:
```diff
@@ -591,6 +592,8 @@ def __init__(
         batch has not been completely filled up.
       retry_strategy: The strategy to use when retrying streaming inserts
         into BigQuery. Options are shown in bigquery_tools.RetryStrategy attrs.
+      additional_bq_parameters (dict, callable): A set of additional parameters
```
Are the names of these parameters unknown, variable ? If not why not add separate keyword arguments for these ?

chamikaramj commented on sdks/python/apache_beam/io/gcp/bigquery.py:
```diff
@@ -855,6 +869,15 @@ def __init__(self,
+        into a table. These can be 'timePartitioning', 'clustering', etc. They
+        are passed directly to the job load configuration. See
+        https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#configuration.load
+      table_side_inputs (tuple): A tuple with ``AsSideInput`` PCollections to be
```
What is "table callable" and "schema callable" ?

chamikaramj commented on sdks/python/apache_beam/io/gcp/bigquery.py:
```diff
@@ -591,6 +592,8 @@ def __init__(
       retry_strategy: The strategy to use when retrying streaming inserts
         into BigQuery. Options are shown in bigquery_tools.RetryStrategy attrs.
+      additional_bq_parameters (dict, callable): A set of additional parameters
+        to be passed when creating a BigQuery table.
```
Also, prob. explain what users are expected to pass here and how they will be used. It's not clear to an end user how this parameter should be constructed and how this will be used by reading the current doc string.

pabloem commented:
Run Python PreCommit

pabloem commented:
@chamikaramj I've improved the top-level documentation for the class. Hopefully that'll clarify some of your questions.

I did not add individual parameters, because I did not want to add 20 parameters to the transform (and new ones every time BQ adds something else).

chamikaramj commented:
Lemme know when this is ready for another look. I think this is mostly done other than few comments I mentioned in the last round.

pabloem commented on sdks/python/apache_beam/io/gcp/bigquery.py:
```diff
@@ -78,15 +78,120 @@
+When creating a new BigQuery table, there are a number of extra parameters
+that one may need to specify. For example, clustering, partitioning, data
+encoding, etc. It is possible to provide these additional parameters by
+passing a Python dictionary as `additional_bq_parameters` to the transform.
```
Pasting my chat response here:

hey cham so for BQ, the reason that  I'm admitting a variable set of properties is that in Java, the BQ sink API did not adequately support users to provide parameters for clustering
which is a newer feature
so when BQ added it, and users wanted to rely on it, they couldnt. And in fact, their pipelines would fail whenever we needed to run multiple loads for the same table with temporary tables
and they had clustered tables
so I decided to accept a dictionary that would make it easy to accept new elements 
but I agree with you it's risky. We could have a whitelist of properties that we admit. WDYT?

chamikaramj replied:
LGTM. But please make sure that Java and Python APIs are consistent in this regard.

pabloem commented:
Run Python PostCommit

pabloem commented:
I've addressed all comments, except for the one relating `additional_bq_parameters`. The question remains open: WE could whitelist a set of admissible parameters (in this case, `['clustering', 'partitioning']`). Thoughts?

pabloem commented:
Run Python PreCommit

pabloem commented:
Passing postcommits:
https://builds.apache.org/job/beam_PostCommit_Python_Verify_PR/636/
https://builds.apache.org/job/beam_PostCommit_Python3_Verify_PR/273/

chamikaramj commented:
Please squash commits and self merge.
